Applications/Utils/ModelPreparation/PartitionMesh/NodeWiseMeshPartitioner.cpp

```diff
@@ -0,0 +1,515 @@
+            {
+                if (_nodes_partition_ids[elem->getNodeIndex(i)] == part_id)
+                {
+                    non_ghost_node_number++;
+                }
+            }
+
+            if (non_ghost_node_number == 0)
+                continue;
+
+            if (non_ghost_node_number == elem->getNumberOfNodes())
+            {
+                partition.regular_elements.push_back(elem);
+                _elements_status[elem_id] = true;
+            }
+            else
+            {
+                partition.ghost_elements.push_back(elem);
+            }
+        }
+
+        // Find the ghost nodes of this partition
+        std::vector<bool> nodes_reserved(_mesh->getNumberOfNodes(), false);
+        for (const auto* ghost_elem : partition.ghost_elements)
+        {
+            for (unsigned i = 0; i < ghost_elem->getNumberOfNodes(); i++)
+            {
+                const unsigned node_id = ghost_elem->getNodeIndex(i);
+                if (nodes_reserved[node_id])
+                    continue;
+
+                if (_nodes_partition_ids[node_id] != part_id)
+                {
+                    if (is_mixed_hl_elem)
+                    {
+                        if (node_id < _mesh->getNumberOfBaseNodes())
+                            partition.nodes.push_back(nodes[node_id]);
+                        else
+                            extra_nodes.push_back(nodes[node_id]);
+                    }
+                    else
+                    {
+                        partition.nodes.push_back(nodes[node_id]);
+                    }
+                    nodes_reserved[node_id] = true;
+                }
+            }
+        }
+        partition.number_of_base_nodes = partition.nodes.size();
+
+        if (is_mixed_hl_elem)
+            partition.nodes.insert(partition.nodes.end(), extra_nodes.begin(),
+                                   extra_nodes.end());
+    }
+
+    renumberNodeIndices();
+}
+
+void NodeWiseMeshPartitioner::renumberNodeIndices()
+{
+    std::size_t node_global_id_offset = 0;
+    for (auto& partition : _partitions)
+    {
+        // Renumber the global indecies.
+        // -- Base nodes
+        for (std::size_t i = 0; i < partition.number_of_non_ghost_base_nodes;
+             i++)
+        {
+            _nodes_global_ids[partition.nodes[i]->getID()] =
+                node_global_id_offset;
+            node_global_id_offset++;
+        }
+        // -- Nodes for high order elements.
+        const std::size_t end_id = partition.number_of_base_nodes +
+                                   partition.number_of_non_ghost_nodes -
+                                   partition.number_of_non_ghost_base_nodes;
+        for (std::size_t i = partition.number_of_base_nodes; i < end_id; i++)
+        {
+            _nodes_global_ids[partition.nodes[i]->getID()] =
+                node_global_id_offset;
+            node_global_id_offset++;
+        }
```

Comment:
Just a question: Both for-loops are identical but for their ranges; Is it possible to make one loop running over the base nodes as well as the nodes for higher order elements?